### ISSUE TITLE
[NTI] Spurious warnings when returning a constructor

### ISSUE THREAD

**Issue by mgiuffrida** [google/closure-compiler#2259]
A couple of problems when trying to use a function that returns a constructor:

1. "Cannot call non-function type" warning when invoking constructor with `new`

```
java -jar compiler.jar test.js --language_in ECMASCRIPT6_STRICT --new_type_inf --checks_only

test.js:9: WARNING - Cannot call non-function type function(new:Object,*=):Object
let f = new myFn();
        ^^^^^^^^^^

0 error(s), 1 warning(s), 68.7% typed
```

```js
function getFn(n) {
  if (n == 1)
    return n > 10 ? Array : Object;
  else
    return Array;
}

const myFn = getFn(42);
let f = new myFn();
```

2. "In found type, property prototype is optional but should be required."

Same code as above, but this time the call is wrapped in its own function.

```
test.js:9: WARNING - Function called in context that expects incompatible type.
Expected : TOP_FUNCTION
Found    : function(new:Array<?>|Number,*=):Array<?>|number
More details:
In found type, property prototype is optional but should be required.
  const myFn = getFn(11);
               ^^^^^^^^^

0 error(s), 1 warning(s), 70.5% typed
```

```js
function getFn(n) {
  if (n == 1)
    return n > 10 ? Array : Number;
  else
    return Array;
}

function blah() {
  const myFn = getFn(11);
  let f = new myFn();
}
```

These problems go away if `getFn` is annotated as `@return {!function(new:Object)}`.

(The code that originally triggered this was more complex, returned custom classes, and so forth.)

v20161201

**Comment by dimvar** [comment]
Thanks for the report. That's a pretty weird bug. In the first example, if I simplify the THEN branch to `return Object`, or if I replace the body of getFn with the THEN branch, I get no warnings. It needs both the IF and the hook expression for the warning to happen.
